compiler/rustc_middle/src/mir/interpret/mod.rs

```diff
@@ -149,15 +131,28 @@ pub use self::error::{
     ScalarSizeMismatch, UndefinedBehaviorInfo, UnsupportedOpInfo, ValidationErrorInfo,
     ValidationErrorKind,
 };
-
+pub use self::pointer::{CtfeProvenance, Pointer, PointerArithmetic, Provenance};
 pub use self::value::Scalar;
 
-pub use self::allocation::{
-    alloc_range, AllocBytes, AllocError, AllocRange, AllocResult, Allocation, ConstAllocation,
-    InitChunk, InitChunkIter,
-};
-
-pub use self::pointer::{CtfeProvenance, Pointer, PointerArithmetic, Provenance};
+use crate::mir;
+use crate::ty::codec::{TyDecoder, TyEncoder};
+use crate::ty::GenericArgKind;
+use crate::ty::{self, Instance, Ty, TyCtxt};
+use rustc_ast::LitKind;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::sync::{HashMapExt, Lock};
+use rustc_errors::ErrorGuaranteed;
+use rustc_hir::def_id::{DefId, LocalDefId};
+use rustc_macros::{HashStable, TyDecodable, TyEncodable, TypeFoldable, TypeVisitable};
+use rustc_middle::ty::print::with_no_trimmed_paths;
+use rustc_serialize::{Decodable, Encodable};
+use rustc_target::abi::{AddressSpace, Endian, HasDataLayout};
+use smallvec::{smallvec, SmallVec};
+use std::fmt;
+use std::io;
+use std::io::{Read, Write};
+use std::num::NonZero;
+use std::sync::atomic::{AtomicU32, Ordering};
```

Comment:
I would prefer if you did not reorder all the imports here.

if you must reorder them, please follow the style of: std, non-rustc dependencies, rustc dependencies, current crate.

Reply:
Could we just set rustfmt [`group_imports`](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#group_imports) to `StdExternalCrate` somewhere?

Reply:
Maybe... though that will mix rustc_ crate imports with fully external imports, it seems? Or can rustfmt take into account which crates are part of the local workspace?

Reply:
Also https://github.com/rust-lang/rustfmt/issues/4709 makes this a complete no-go I think.

Reply:
> If you must reorder them, please follow the style of: std, non-rustc dependencies, rustc dependencies, current crate.

I estimate that approximately zero files in the codebase use that ordering. There's unfortunately not a lot of consistency about how `use` items are sectioned and ordered, though "put them all in one alphabetical section" is dead simple, possibly the most common, how I usually do it, and nobody has ever complained about it before.

Reply:
I use that ordering in files I maintain.

I ask you not mess with the order of other files when editing parts of the compiler for which you are not the primary maintainer.

Reply:
> I estimate that approximately zero files in the codebase use that ordering.

Counterexamples: [1](https://github.com/rust-lang/rust/blob/875f0c2da05da9a7620afd6e2be04fbcb9a4b395/compiler/rustc_const_eval/src/interpret/place.rs#L5-L14), [2](https://github.com/rust-lang/rust/blob/4497990dff8ce181fa12757eb4e786b85025ae61/compiler/rustc_const_eval/src/interpret/projection.rs#L10-L20), [3](https://github.com/rust-lang/rust/blob/bb8d6f790b63797a816a4007e8ab002ed2906de7/compiler/rustc_const_eval/src/interpret/operand.rs#L4-L14), [4](https://github.com/rust-lang/rust/blob/ae7b07f2dc9b07aceab318829538222666945efc/compiler/rustc_const_eval/src/interpret/memory.rs#L9-L29), [5](https://github.com/rust-lang/rust/blob/126dcc618d065e97887113926f63986bd0a9e681/compiler/rustc_const_eval/src/interpret/intern.rs#L16-L29).

I think about 80% of the files in the `interpret` folder follow that ordering (and some of the ones that don't still almost do, they may just have a `crate` import separate from and *after* a `self` import which is a bit odd).

Reply:
I have reverted that part of the change. That file is strange, with macros and `mod` items before the `use` items.

It is unfortunate that rustfmt is so unopinionated about `use` item ordering, and that `group_imports` is not mature enough to be usable.

> I ask you not apply your own style for the order of imports when editing parts of the compiler for which you are not the primary maintainer.

I'm going to push back strongly here. I have no problem with people arguing against changes on the merits, but this sentence goes well beyond that. "Primary maintainer" is not something that has any official meaning within the Rust project. Nor does the project have any principle that suggests a commit X would be acceptable if done by person A but unacceptable if done by person B. Please don't invent process in an attempt to block a change.

Reply:
I should have worded this better.  We don't have a formal concept of "primary maintainer", agreed, but we have de-facto people focusing on particular parts of the compiler, and being responsible for design decisions affecting that part. Larger interpreter changes will involve wg-const-eval. Formatting changes are not quite the same, but we also do have a policy against formatting-only changes, except when accompanied by tool support (and, where appropriate, backed by team consensus). This was established in reply to people making sweeping PRs that did nothing but add trailing `.` to comments or things like that. I don't know where we document such policy decisions.

I agree the file is strange, and I wouldn't be opposed to some sort of reordering that doesn't change the formatting of `use` -- TBH I never found a good style for how to order `mod` and `use`. IIRC the macros have to (or had to) go first because of macro magic; not sure if that still applies.

Thanks for reverting the changes.